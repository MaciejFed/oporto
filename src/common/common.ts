import { DateTime } from 'luxon';
import { Zone } from 'luxon/src/zone';
import { DurationLike } from 'luxon/src/duration';
import { Exercise } from '../exercise/exercise';

export class DateTimeExtended {
  private readonly date: DateTime;

  private constructor(date: Date) {
    this.date = DateTime.fromJSDate(date);
  }

  public plus(duration: DurationLike) {
    const newDate = this.date.plus(duration);
    return DateTimeExtended.fromJSDate(newDate.toJSDate());
  }

  public toJSDate(): Date {
    return this.date.toJSDate();
  }

  get ordinal(): number {
    return this.date.year * 365 + this.date.ordinal;
  }

  get weekday(): number {
    return this.date.weekday;
  }

  get weekNumber(): number {
    this.date.toJSDate();
    return this.date.weekNumber;
  }

  static fromJSDate(date: Date, options?: { zone?: string | Zone }): DateTimeExtended {
    return new DateTimeExtended(date);
  }

  static now(): DateTimeExtended {
    return new DateTimeExtended(new Date());
  }
}

export interface Comparable {
  equal(other: Comparable): boolean;
}

export function onlyDistinct(arr: Comparable[]): Comparable[] {
  const distinctElements: Comparable[] = [];
  for (let i = 0; i < arr.length; i++) {
    let unique = true;
    for (let j = 0; j < distinctElements.length; j++) {
      if (i !== j && arr[i].equal(distinctElements[j])) {
        unique = false;
      }
    }
    if (unique) {
      distinctElements.push(arr[i]);
    }
  }
  return distinctElements;
}

export function enforceArrayLimit(arr: any[], limit: number) {
  if (arr.length > limit) {
    arr.splice(limit);
  }
  return arr;
}

export function onlyDistinctSubjects(arr: Exercise[]): Exercise[] {
  const distinctElements: Exercise[] = [];
  for (let i = 0; i < arr.length; i++) {
    let unique = true;
    for (let j = 0; j < distinctElements.length; j++) {
      if (i !== j && arr[i].getBaseWord() === distinctElements[j].getBaseWord()) {
        unique = false;
      }
    }
    if (unique) {
      distinctElements.push(arr[i]);
    }
  }
  return distinctElements;
}

export function findWordStartIndex(sentence: string, word: string) {
  const words = sentence.split(' ');
  const index = words.indexOf(word);

  return words.slice(0, index).reduce((prev, curr) => curr.length + prev + 1, 0);
}

export function getRandomElement<T>(arr: T[], maxLength?: number): T {
  // eslint-disable-next-line no-nested-ternary
  const length = maxLength ? (arr.length > maxLength ? maxLength : arr.length) : arr.length;
  return arr[Math.floor(Math.random() * length)];
}

export async function sleep(milliseconds: number) {
  await new Promise((resolve) => setTimeout(resolve, milliseconds));
}

export function formatDate(date: Date): string {
  const dateTime = DateTime.fromJSDate(date);
  return `${date.getFullYear()}-${
    date.getMonth() + 1
  }-${date.getDate()} ${date.getHours()}:${date.getMinutes()} [${dateTime.toRelative()}]`;
}

export function isBeforeWeekday(resultDate: Date, weekday: number) {
  const now = DateTime.now();
  const dayInYearNow = (now.weekNumber - 1) * 7 + weekday;

  const resultDateTime = DateTime.fromJSDate(resultDate);
  const resultDayInYear = (resultDateTime.weekNumber - 1) * 7 + resultDateTime.weekday;

  return resultDayInYear <= dayInYearNow;
}

export function isOnWeekDay(resultDate: Date, weekday: number, year: number) {
  const now = DateTime.now();
  const resultDateTime = DateTime.fromJSDate(resultDate);

  return (
    now.weekNumber === resultDateTime.weekNumber && resultDateTime.weekday === weekday && resultDateTime.year === year
  );
}

export function areSameCalendarDay(date1: Date, date2: Date): boolean {
  return (
    date1.getFullYear() === date2.getFullYear() &&
    date1.getMonth() === date2.getMonth() &&
    date1.getDate() === date2.getDate()
  );
}

export function assertNonNull<T>(value?: T): T {
  if (!value) throw new Error('Unexpected Undefined!');

  return value;
}
