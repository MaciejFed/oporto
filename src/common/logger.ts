import winston from 'winston';

const winstonLogger = winston.createLogger({
  level: 'info',
  format: winston.format.combine(
    winston.format.timestamp({ format: 'YYYY-MM-DD HH:mm:ss' }),
    winston.format.prettyPrint()
  ),
  transports: [new winston.transports.File({ filename: 'out.log' })]
});

export const logger = {
  info: (msg: string, ...arg: any[]) => {
    console.log(msg, arg);
    winstonLogger.info(msg);
  },
  debug: (msg: string, ...arg: any[]) => {
    console.debug(msg, arg);
    winstonLogger.debug(msg);
  },
  error: (msg: string, ...arg: any[]) => {
    console.error(msg, arg);
    winstonLogger.error(msg, arg);
  }
}
