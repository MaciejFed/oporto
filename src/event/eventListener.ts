export interface AppEventListener {
    registerListeners(): void
}