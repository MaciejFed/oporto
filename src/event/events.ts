export const APP_STARTED = 'APP_STARTED';
export const APP_FINISHED = 'APP_FINISHED';
export const KEY_PRESSED = 'KEY_PRESSED';
export const ANSWER_SUBMITTED = 'ANSWER_SUBMITTED';
export const ANSWER_CHECKED = 'ANSWER_CHECKED';
export const NEW_WORD_LEARNED = 'NEW_WORD_LEARNED';
export const EXERCISE_STARTED = 'EXERCISE_STARTED';
export const EXERCISE_NEXT = 'EXERCISE_NEXT';
export const EXERCISE_DESCRIPTION_PRINTED = 'EXERCISE_DESCRIPTION_PRINTED';
export const EXERCISE_BODY_PRINTED = 'EXERCISE_BODY_PRINTED';
export const TERMINAL_CLEARED = 'TERMINAL_CLEARED';
export const HEARING_EXERCISE_REPEAT = 'HEARING_EXERCISE_REPEAT';

export type EXERCISE_BODY_PRINTED_BODY = {
  exerciseBodyPrefix: string;
  exerciseBodySuffix: string;
  exerciseTranslation: string;
};

export type APP_EVENT =
  | typeof APP_STARTED
  | typeof KEY_PRESSED
  | typeof ANSWER_SUBMITTED
  | typeof ANSWER_CHECKED
  | typeof EXERCISE_STARTED
  | typeof EXERCISE_NEXT
  | typeof EXERCISE_DESCRIPTION_PRINTED
  | typeof NEW_WORD_LEARNED
  | typeof EXERCISE_BODY_PRINTED
  | typeof APP_FINISHED
  | typeof TERMINAL_CLEARED
  | typeof HEARING_EXERCISE_REPEAT;
