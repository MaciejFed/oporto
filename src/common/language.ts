import { getRandomElement } from './common';

export enum Language {
  Portuguese = 'Portuguese',
  German = 'German'
}

export function getVoice(language: Language): string {
  if (language === Language.Portuguese) {
    return getRandomElement(['Joana', 'Joaquim']);
  }
  return getRandomElement(['Anna', 'Petra', 'Markus']);
}
