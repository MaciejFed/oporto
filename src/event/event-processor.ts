import EventEmitter from 'events';
import { terminal } from 'terminal-kit';
import { displayStatistics } from '../commands/stat';
import { logger } from '../common/logger';
import { APP_EVENT } from './events';
import Output from '../io/output';

type EmittedEvent = {
  event: APP_EVENT;
  args: any[];
};

export class EventProcessor {
  eventEmitter: EventEmitter;
  eventHistory: EmittedEvent[];

  constructor() {
    this.eventEmitter = new EventEmitter();
    this.eventHistory = [];
  }

  public on(eventName: APP_EVENT, callback: (...args: any[]) => void) {
    this.eventEmitter.on(eventName, callback);
  }

  public emit(event: APP_EVENT, args?: any) {
    this.eventHistory.push({
      event: event,
      args: args
    });
    this.logEvent(event, args);
    this.doEmitEvent(event, args);
  }

  private logEvent(event: APP_EVENT, args?: any) {
    if (args && typeof args === 'object') {
      logger.debug(`${event}: [${JSON.stringify(args)}]`);
    } else {
      logger.debug(`${event}: [${args}]`);
    }
  }

  private doEmitEvent(event: APP_EVENT, args?: any) {
    switch (event) {
      case 'APP_FINISHED':
        this.eventEmitter.removeAllListeners();
        terminal.hideCursor(false);
        displayStatistics(false);
        process.exit(0);
        break;
      default:
        this.eventEmitter.emit(event, args);
    }
  }
}

export default new EventProcessor();
