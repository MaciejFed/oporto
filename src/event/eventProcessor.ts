import clear from 'clear';
import EventEmitter from 'events';
import { terminal } from 'terminal-kit';
import winston from 'winston';
import { displayStatistics } from '../commands/stat';
import { logger } from '../common/logger';
import { APP_EVENT, APP_STARTED } from './events';

type EmittedEvent = {
  event: APP_EVENT;
  args: any[];
};

export class EventProcessor {
  eventEmitter: EventEmitter;
  eventHistory: EmittedEvent[];

  constructor(eventEmitter: EventEmitter) {
    this.eventEmitter = eventEmitter;
    this.eventHistory = [];
  }

  on(eventName: string, callback: (...args: any[]) => void) {
    this.eventEmitter.on(eventName, callback);
  }

  emit(event: APP_EVENT, args?: any) {
    this.eventHistory.push({
      event: event,
      args: args
    });
    this.logEvent(event, args);
    this.doEmitEvent(event, args);
  }

  private logEvent(event: APP_EVENT, args?: any) {
    if (args && typeof args === 'object') {
      logger.info(`${event}: [${JSON.stringify(args)}]`);
    } else {
      logger.info(`${event}: [${args}]`);
    }
  }

  private doEmitEvent(event: APP_EVENT, args?: any) {
    switch (event) {
      case 'APP_FINISHED':
        this.eventEmitter.removeAllListeners();
        terminal.hideCursor(false);
        clear();
        displayStatistics(false);
        process.exit(0);
        break;
      default:
        this.eventEmitter.emit(event, args);
    }
  }
}

export const eventProcessor = new EventProcessor(new EventEmitter());
