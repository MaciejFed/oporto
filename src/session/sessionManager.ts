import {
  ANSWER_SUBMITED,
  ANSWER_CHECKED,
  APP_FINISHED,
  APP_STARTED,
  EXERCISE_BODY_PRINTED,
  EXERCISE_DESCRIPTION_PRINTED,
  EXERCISE_NEXT,
  EXERCISE_STARTED,
  KEY_PRESSED
} from '../event/events';
import { AppEventListener } from '../event/eventListener';
import { EventProcessor } from '../event/eventProcessor';
import { Exercise, generateUniqeExercises } from '../exercise/exercise';
import { logger } from '../logger/logger';
import { convertToResult, Result } from '../service/result';
import { saveNewResult } from '../repository/resultRepository';
import { AnswerInputType } from '../input/input';

export class SessionManager implements AppEventListener {
  eventProcessor: EventProcessor;
  exercises: Exercise[];
  results: Result[];
  currentExercise: Exercise;
  answer: string;
  exerciseInProgress: boolean;

  constructor(eventProcessor: EventProcessor, exerciseCount: number) {
    this.eventProcessor = eventProcessor;
    this.registerListeners();
    this.exercises = generateUniqeExercises(exerciseCount);
    this.results = [];
    this.currentExercise = this.exercises[0];
    this.answer = '';
    this.exerciseInProgress = false;
  }

  registerListeners() {
    this.registerAppStartedEventListener();
    this.registerExerciseStartedEventListener();
    this.registerKeyPressedEventListener();
    this.registerAnswerSubmittedEventListener();
    this.registerNextExerciseEventListener();
  }

  registerAppStartedEventListener() {
    this.eventProcessor.on(APP_STARTED, () => {
      this.eventProcessor.emit(EXERCISE_STARTED);
    });
  }

  registerExerciseStartedEventListener() {
    this.eventProcessor.on(EXERCISE_STARTED, () => {
      this.currentExercise = this.exercises.pop() || this.exercises[0];
      this.eventProcessor.emit(EXERCISE_DESCRIPTION_PRINTED, this.currentExercise?.getExerciseDescription());
      this.eventProcessor.emit(EXERCISE_BODY_PRINTED, {
        exercise: this.currentExercise?.getExerciseBody(),
        cursor: {
          x: 0,
          y: 0
        }
      });
      this.exerciseInProgress = true;
    });
  }

  registerKeyPressedEventListener() {
    this.eventProcessor.on(KEY_PRESSED, (key) => {
      if (key === 'backspace') {
        this.answer = this.answer.substring(0, Math.max(0, this.answer.length - 1));
      } else {
        this.answer += key;
      }
      logger.info(`Answer: "${this.answer}"`);
    });
  }

  registerAnswerSubmittedEventListener() {
    this.eventProcessor.on(ANSWER_SUBMITED, (answerInputType: AnswerInputType) => {
      if (!this.exerciseInProgress) {
        return;
      }
      this.exerciseInProgress = false;
      const correctAnswer = this.currentExercise?.getCorrectAnswer().toLowerCase();
      const isCorrect = this.currentExercise?.checkAnsweCorrect(this.answer);
      saveNewResult(convertToResult(this.currentExercise, this.answer, answerInputType));
      logger.info(`Answer: "${this.answer}", correctAnswer: "${correctAnswer}" `);
      this.eventProcessor.emit(ANSWER_CHECKED, {
        isCorrect,
        correctAnswer,
        answerInputType
      });
      this.resetAnswer();
    });
  }

  registerNextExerciseEventListener() {
    this.eventProcessor.on(EXERCISE_NEXT, () => {
      this.resetAnswer();
      if (this.exercises.length > 0) {
        this.eventProcessor.emit(EXERCISE_STARTED);
      } else {
        process.stdin.removeAllListeners();
        this.eventProcessor.emit(APP_FINISHED);
      }
    });
  }

  resetAnswer() {
    logger.info('Reseting answer...');
    this.answer = '';
  }
}
