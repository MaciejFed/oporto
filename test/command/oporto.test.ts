/* eslint-disable @typescript-eslint/no-var-requires */
import { sleep } from '../../src/common/common';
import { simulateTyping } from '../util';

const output: string[] = [];
const resultCount = {
  greenCount: 0,
  redCount: 0
};
const sayCommands: string[] = [];

 // @ts-ignore
 process.stdin.setRawMode = () => {};

jest.mock('child_process', () => {
  return {
    __esModule: true,
    exec: (command: string) => { sayCommands.push(command) }
  };
});
jest.mock('terminal-kit', () => {
    return {
      __esModule: true,
      terminal: {
        moveTo: (x: number, y: number, data: string) => {
          if (data)
            output.push(data);
        },
        hideCursor: (hide: boolean) => {},
        green: () => {
          resultCount.greenCount++;
        },
        red: () => {
          resultCount.redCount++;
        },
        white: () => {},
        bold: () => {},
      }
    };
  });
jest.mock('../../src/service/verb', () => {
  const modeuleActual = jest.requireActual('../../src/service/verb');
  const readAll = require('../../src/repository/exercisesRepository').readAll;
  return {
    ...modeuleActual,
    getRandomRegularVerb: () => Object.assign({ portuguese: 'Comer', english: '' }),
    getRandomPerson: () => 'Eu',
    getRandomIrregularVerb: () => readAll().verbs.irregular[0],
  };
});

jest.mock('../../src/exercise/exercise', () => {
  const RegularVerbExercise = require('../../src/exercise/verbExercise').RegularVerbExercise;
  const modeuleActual = jest.requireActual('../../src/exercise/exercise');
  return {
    __esModule: true,
    modeuleActual,
    generateUniqeExercises: () => [new RegularVerbExercise()]
  };
});


type AppModules = {
  SessionManager: any;
  Input: any;
  Terminakl: any;
  eventProcessor: any;
};


function requireAllModules(): AppModules {
  const eventProcessor =
    require('../../src/event/eventProcessor').eventProcessor;
  const SessionManager =
    require('../../src/session/sessionManager').SessionManager;
  const Input = require('../../src/io/input').Input;
  const Terminakl = require('../../src/io/terminal').Terminakl;

  return {
    SessionManager: SessionManager,
    Input: Input,
    Terminakl: Terminakl,
    eventProcessor: eventProcessor
  };
}

describe('Event Emitter', () => {
  beforeEach(() => {
    // @ts-ignore
    const mockExit = jest.spyOn(process, 'exit').mockImplementation(() => {});
    sayCommands.length = 0;
    output.length = 0;
    resultCount.greenCount = 0;
    resultCount.redCount = 0;

  });

  afterEach(() => {
    jest.resetModules();
    process.stdin.removeAllListeners();
  });

  it('Happy Path', async () => {
    const appModules = requireAllModules();
    const myTerminal = new appModules.Terminakl(appModules.eventProcessor);
    const input = new appModules.Input(appModules.eventProcessor);
    const sessionManager = new appModules.SessionManager(
      appModules.eventProcessor,
      1
    );
    appModules.eventProcessor.emit('APP_STARTED');
    console.log(process.stdin.listeners);
    simulateTyping('como');
    process.stdin.emit('keypress', {}, { name: 'return' });
    process.stdin.emit('keypress', {}, { name: 'n' });

    await sleep(50);

    const eventHistory = appModules.eventProcessor.eventHistory.map(
      (event: { event: any }) => event.event
    );

    expect(output[0]).toEqual('Infinitive: Comer');
    expect(output[5]).toEqual('Eu: como');
    expect(output[6]).toEqual('Correct! [voice]');
    expect(resultCount.greenCount).toBe('como'.length);
    expect(resultCount.redCount).toBe(0);
    expect(sayCommands[0]).toEqual('say "Eu Como"');
    expect(eventHistory).toMatchSnapshot();
    expect(output).toMatchSnapshot();
  });

  it('Unhappy Path', async () => {
    console.log(process.stdin.listeners);
    const appModules = requireAllModules();
    const myTerminal = new appModules.Terminakl(appModules.eventProcessor);
    const input = new appModules.Input(appModules.eventProcessor);
    const sessionManager = new appModules.SessionManager(
      appModules.eventProcessor,
      1
    );

    appModules.eventProcessor.emit('APP_STARTED');
    simulateTyping('cowronganswer');
    process.stdin.emit('keypress', {}, { name: 'return' });

    await sleep(50);

    const eventHistory = appModules.eventProcessor.eventHistory.map(
        (event: { event: any }) => event.event
      );

    expect(output[0]).toEqual('Infinitive: Comer');
    expect(output[14]).toEqual('Eu: cowronganswer');
    expect(output[15]).toEqual('Wrong! [voice]');
    expect(resultCount.greenCount).toBe('co'.length);
    expect(resultCount.redCount).toBe(2);
    expect(sayCommands[0]).toEqual('say "Eu Como"');
    expect(eventHistory).toMatchSnapshot();
    expect(output).toMatchSnapshot();
  });

  it('HP With Backspace', async () => {
    console.log(process.stdin.listeners);
    const appModules = requireAllModules();
    const myTerminal = new appModules.Terminakl(appModules.eventProcessor);
    const input = new appModules.Input(appModules.eventProcessor);
    const sessionManager = new appModules.SessionManager(
      appModules.eventProcessor,
      1
    );
    appModules.eventProcessor.emit('APP_STARTED');
    simulateTyping('comorr');
    process.stdin.emit('keypress', {}, { name: 'backspace' });
    process.stdin.emit('keypress', {}, { name: 'backspace' });
    process.stdin.emit('keypress', {}, { name: 'return' });
    process.stdin.emit('keypress', {}, { name: 'n' });

    await sleep(50);
    
    const eventHistory = appModules.eventProcessor.eventHistory.map(
        (event: { event: any }) => event.event
      );

    expect(output[0]).toEqual('Infinitive: Comer');
    expect(output[7]).toEqual('Eu: comorr');
    expect(output[9]).toEqual('Eu: como');
    expect(output[10]).toEqual('Correct! [voice]');
    expect(resultCount.greenCount).toBe('como'.length);
    expect(resultCount.redCount).toBe(0);
    expect(sayCommands[0]).toEqual('say "Eu Como"');
    expect(output).toMatchSnapshot();
    expect(eventHistory).toMatchSnapshot();
  });
});
