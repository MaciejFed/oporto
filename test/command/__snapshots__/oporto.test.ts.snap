// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Event Emitter HP With Backspace 1`] = `
Array [
  "Infinitive: to be - long",
  "Eu: ",
  "Eu: s",
  "Eu: so",
  "Eu: sou",
  "Eu: sour",
  "Eu: sourr",
  "Eu: sour",
  "Eu: sou",
  "Correct! [voice]",
  "Eu: ",
  "s",
  "o",
  "u",
  "Press key to continue...",
  "r - repeat the answer",
  "Weekly Statistics:

",
  "Weekly Statistics:

",
]
`;

exports[`Event Emitter HP With Backspace 2`] = `
Array [
  "APP_STARTED",
  "EXERCISE_STARTED",
  "EXERCISE_DESCRIPTION_PRINTED",
  "EXERCISE_BODY_PRINTED",
  "KEY_PRESSSED",
  "KEY_PRESSSED",
  "KEY_PRESSSED",
  "KEY_PRESSSED",
  "KEY_PRESSSED",
  "KEY_PRESSSED",
  "KEY_PRESSSED",
  "ANSWER_SUBMITED",
  "ANSWER_CHECKED",
  "KEY_PRESSSED",
  "EXERCISE_NEXT",
  "APP_FINISHED",
]
`;

exports[`Event Emitter Happy Path 1`] = `
Array [
  "APP_STARTED",
  "EXERCISE_STARTED",
  "EXERCISE_DESCRIPTION_PRINTED",
  "EXERCISE_BODY_PRINTED",
  "KEY_PRESSSED",
  "KEY_PRESSSED",
  "KEY_PRESSSED",
  "ANSWER_SUBMITED",
  "ANSWER_CHECKED",
  "KEY_PRESSSED",
  "EXERCISE_NEXT",
  "APP_FINISHED",
]
`;

exports[`Event Emitter Happy Path 2`] = `
Array [
  "Infinitive: to be - long",
  "Eu: ",
  "Eu: s",
  "Eu: so",
  "Eu: sou",
  "Correct! [voice]",
  "Eu: ",
  "s",
  "o",
  "u",
  "Press key to continue...",
  "r - repeat the answer",
  "Weekly Statistics:

",
  "Weekly Statistics:

",
]
`;

exports[`Event Emitter Unhappy Path 1`] = `
Array [
  "APP_STARTED",
  "EXERCISE_STARTED",
  "EXERCISE_DESCRIPTION_PRINTED",
  "EXERCISE_BODY_PRINTED",
  "KEY_PRESSSED",
  "KEY_PRESSSED",
  "KEY_PRESSSED",
  "KEY_PRESSSED",
  "KEY_PRESSSED",
  "KEY_PRESSSED",
  "KEY_PRESSSED",
  "KEY_PRESSSED",
  "KEY_PRESSSED",
  "KEY_PRESSSED",
  "KEY_PRESSSED",
  "KEY_PRESSSED",
  "KEY_PRESSSED",
  "ANSWER_SUBMITED",
  "ANSWER_CHECKED",
]
`;

exports[`Event Emitter Unhappy Path 2`] = `
Array [
  "Infinitive: to be - long",
  "Eu: ",
  "Eu: s",
  "Eu: so",
  "Eu: sow",
  "Eu: sowr",
  "Eu: sowro",
  "Eu: sowron",
  "Eu: sowrong",
  "Eu: sowronga",
  "Eu: sowrongan",
  "Eu: sowrongans",
  "Eu: sowrongansw",
  "Eu: sowronganswe",
  "Eu: sowronganswer",
  "Wrong! [voice]",
  "Eu: ",
  "s",
  "o",
  "u",
  "Repeat: ",
]
`;
